modules/repofiles/action.go

```diff
@@ -29,187 +29,209 @@ type CommitRepoActionOptions struct {
+				Repo:       apiRepo,
+				Sender:     apiPusher,
+			}); err != nil {
+				return fmt.Errorf("PrepareWebhooks.(delete tag): %v", err)
+			}
 		}
 
-	case models.ActionDeleteBranch: // Delete Branch
-		isHookEventPush = true
-
-		if err = models.PrepareWebhooks(repo, models.HookEventDelete, &api.DeletePayload{
-			Ref:        refName,
-			RefType:    "branch",
-			PusherType: api.PusherTypeUser,
-			Repo:       apiRepo,
-			Sender:     apiPusher,
-		}); err != nil {
-			return fmt.Errorf("PrepareWebhooks.(delete branch): %v", err)
-		}
-
-	case models.ActionPushTag: // Create
-		isHookEventPush = true
-
-		gitRepo, err := git.OpenRepository(repo.RepoPath())
-		if err != nil {
-			log.Error("OpenRepository[%s]: %v", repo.RepoPath(), err)
-		}
-		shaSum, err = gitRepo.GetTagCommitID(refName)
-		if err != nil {
-			log.Error("GetTagCommitID[%s]: %v", opts.RefFullName, err)
-		}
-		if err = models.PrepareWebhooks(repo, models.HookEventCreate, &api.CreatePayload{
-			Ref:     refName,
-			Sha:     shaSum,
-			RefType: "tag",
-			Repo:    apiRepo,
-			Sender:  apiPusher,
-		}); err != nil {
-			return fmt.Errorf("PrepareWebhooks: %v", err)
-		}
-	case models.ActionDeleteTag: // Delete Tag
-		isHookEventPush = true
-
-		if err = models.PrepareWebhooks(repo, models.HookEventDelete, &api.DeletePayload{
-			Ref:        refName,
-			RefType:    "tag",
-			PusherType: api.PusherTypeUser,
-			Repo:       apiRepo,
-			Sender:     apiPusher,
-		}); err != nil {
-			return fmt.Errorf("PrepareWebhooks.(delete tag): %v", err)
+		if isHookEventPush {
+			commits, err := opts.Commits.ToAPIPayloadCommits(repo.RepoPath(), repo.HTMLURL())
+			if err != nil {
+				return err
+			}
+			if err = models.PrepareWebhooks(repo, models.HookEventPush, &api.PushPayload{
+				Ref:        opts.RefFullName,
+				Before:     opts.OldCommitID,
+				After:      opts.NewCommitID,
+				CompareURL: setting.AppURL + opts.Commits.CompareURL,
+				Commits:    commits,
+				Repo:       apiRepo,
+				Pusher:     apiPusher,
+				Sender:     apiPusher,
+			}); err != nil {
+				return fmt.Errorf("PrepareWebhooks: %v", err)
+			}
 		}
 	}
 
-	if isHookEventPush {
-		commits, err := opts.Commits.ToAPIPayloadCommits(repo.RepoPath(), repo.HTMLURL())
-		if err != nil {
-			return err
-		}
-		if err = models.PrepareWebhooks(repo, models.HookEventPush, &api.PushPayload{
-			Ref:        opts.RefFullName,
-			Before:     opts.OldCommitID,
-			After:      opts.NewCommitID,
-			CompareURL: setting.AppURL + opts.Commits.CompareURL,
-			Commits:    commits,
-			Repo:       apiRepo,
-			Pusher:     apiPusher,
-			Sender:     apiPusher,
-		}); err != nil {
-			return fmt.Errorf("PrepareWebhooks: %v", err)
+	if repo != nil {
+		// Change repository empty status and update last updated time.
+		if err := models.UpdateRepository(repo, false); err != nil {
+			return fmt.Errorf("UpdateRepository: %v", err)
 		}
 	}
 
+	if err := models.NotifyWatchers(actions...); err != nil {
```

Comment:
I wonder if we really need this to happen within the processing goroutine - or whether we could get away with sticking this a in another goroutine.

Reply:
All notification could be a seperate goroutine.

Reply:
See #8874